bluemira/equilibria/run.py

```diff
@@ -0,0 +1,460 @@
+            max_currents = self.coilset.get_max_currents(0)
+
+            problem = self._bd_prob_cls(
+                coilset,
+                strategy,
+                B_stray_max=self.params.B_premag_stray_max.value,
+                B_stray_con_tol=self._bd_settings["B_stray_con_tol"],
+                n_B_stray_points=self._bd_settings["n_B_stray_points"],
+                optimiser=self._bd_opt,
+                max_currents=max_currents,
+                constraints=constraints,
+            )
+            coilset = problem.optimise(max_currents / 1e6)
+            breakdown = Breakdown(coilset, self.grid, R_0=R_0)
+            breakdown.set_breakdown_point(*strategy.breakdown_point)
+            psi_premag = breakdown.breakdown_psi
+            bluemira_print(f"Premagnetisation flux = {2*np.pi * psi_premag:.2f} V.s")
+
+            if i == 0:
+                psi_1 = psi_premag
+            else:
+                relaxed = np.isclose(psi_premag, psi_1, rtol=1e-2)
+                psi_1 = psi_premag
+            i += 1
+
+        self.take_snapshot(self.BREAKDOWN, breakdown, coilset, self._bd_opt, problem)
+
+    def optimise_currents(self):
+        """
+        Optimise the coil currents at the start and end of the current flat-top.
+        """
+        psi_sof, psi_eof = self.calculate_sof_eof_fluxes()
+        if self.EQ_REF not in self.snapshots:
+            self.run_reference_equilibrium()
+        eq_ref = self.snapshots[self.EQ_REF].eq
+
+        snapshots = [self.SOF, self.EOF]
+
+        max_currents = self.coilset.get_max_currents(0)
+        for snap, psi_boundary in zip(snapshots, [psi_sof, psi_eof]):
+            eq = deepcopy(eq_ref)
+            fixed_coils = all([c.flag_sizefix for c in eq.coilset.coils.values()])
+            eq.coilset.mesh_coils(0.2)
+            self.eq_targets(eq, I_not_dI=True, fixed_coils=fixed_coils)
+            self.eq_targets.update_psi_boundary(psi_boundary / (2 * np.pi))
+            _, A, b = self.eq_targets.get_weighted_arrays()
+
+            L2_target_constraint = OptimisationConstraint(  # noqa: N806
+                L2_norm_constraint,
+                f_constraint_args={
+                    "a_mat": A,
+                    "b_vec": b,
+                    "value": self._eq_settings["target_rms_max"],
+                    "scale": 1e6,
+                },
+                tolerance=np.array([self._eq_settings["target_rms_con_tol"]]),
+            )
+
+            optimiser = deepcopy(self._eq_opt)
+            constraints = deepcopy(self._coil_cons)
+            for constraint in constraints:
+                constraint._args["eq"] = eq
+            constraints.append(L2_target_constraint)
+            problem = self._eq_prob_cls(eq, max_currents, optimiser, constraints)
+            coilset = problem.optimise()
+            self.take_snapshot(snap, eq, coilset, optimiser, self.eq_targets, profiles)
+
+
+if __name__ == "__main__":
```

Comment:
Can this whole section be moved to an example or a test? I dont think we should merge it here either way